### DISCUSSION TITLE
how to interpret `tree-sitter parse   --debug` ?

### DISCUSSION THREAD

**Discussion by teto** [tree-sitter/tree-sitter#2075] [Q&A]
I've postponed my treesitter journey since it seemed overwhelming. It kinda is but it is also a fantastic tool and I am loving it. 

I've got plenty of errors in a grammar I write, and I dont necessarily ask for an answer, but I would appreciate some help understanding the`tree-sitter parse tests/external_body.http  --debug` output. I've extracted what I think is the most straightforward error:

I am trying to parse http headers via
```
    header: ($) =>
      seq(
        field("name", seq($.name, ":")),
        $._whitespace,
        field("value", $.value)
      ),
    
    name: (_) => /[A-Za-z0-9_-]+/,
    // value: (_) => /[^\n]+/,
    value: (_) => /\w+/,

```
in the following file
```
POST https://reqres.in/api/users
X-header-0: toto
X-header-1: tata
```
The `tree-sitter parse tests/external_body.http  --debug` output (look for my annotation FIRST ERROR)
```
new_parse
process version:0, version_count:1, state:1, row:0, col:0
lex_internal state:25, row:0, column:0
  consume character:'P'
  consume character:'O'
  consume character:'S'
  consume character:'T'
lexed_lookahead sym:method, size:4
shift state:14
process version:0, version_count:1, state:14, row:0, col:4
lex_internal state:4, row:0, column:4
  consume character:' '
lexed_lookahead sym:_whitespace_token1, size:1
shift state:11
process version:0, version_count:1, state:11, row:0, col:5
lex_internal state:2, row:0, column:5
  consume character:'h'
  consume character:'t'
  consume character:'t'
  consume character:'p'
  consume character:'s'
  consume character:':'
  consume character:'/'
  consume character:'/'
  consume character:'r'
  consume character:'e'
  consume character:'q'
  consume character:'r'
  consume character:'e'
  consume character:'s'
  consume character:'.'
  consume character:'i'
  consume character:'n'
  consume character:'/'
  consume character:'a'
  consume character:'p'
  consume character:'i'
  consume character:'/'
  consume character:'u'
  consume character:'s'
  consume character:'e'
  consume character:'r'
  consume character:'s'
lexed_lookahead sym:url, size:27
shift state:10
process version:0, version_count:1, state:10, row:0, col:32
lex_internal state:3, row:0, column:32
  consume character:10
lexed_lookahead sym:\n, size:1
reduce sym:request, child_count:3
shift state:8
process version:0, version_count:1, state:8, row:1, col:0
lex_internal state:25, row:1, column:0
lex_internal state:0, row:1, column:0
  consume character:'X'
  consume character:'-'
  consume character:'h'
  consume character:'e'
  consume character:'a'
  consume character:'d'
  consume character:'e'
  consume character:'r'
  consume character:'-'
  consume character:'0'
lexed_lookahead sym:name, size:10
detect_error                      <-----------------------------    FIRST ERROR after parsing "X-header-0", it reparses it again successfully 
resume version:0
recover_to_previous state:7, depth:2
skip_token symbol:name
process version:0, version_count:2, state:0, row:1, col:10
lex_internal state:0, row:1, column:10
  consume character:':'
lexed_lookahead sym::, size:1
skip_token symbol::
process version:1, version_count:2, state:7, row:1, col:0
lex_internal state:1, row:1, column:0
  consume character:'X'
  consume character:'-'
  consume character:'h'
  consume character:'e'
  consume character:'a'
  consume character:'d'
  consume character:'e'
  consume character:'r'
  consume character:'-'
  consume character:'0'
lexed_lookahead sym:name, size:10
shift state:21
process version:1, version_count:2, state:21, row:1, col:10
lex_internal state:0, row:1, column:10
  consume character:':'
lexed_lookahead sym::, size:1
shift state:16
condense
process version:0, version_count:1, state:16, row:1, col:11
lex_internal state:4, row:1, column:11
  consume character:' '
lexed_lookahead sym:_whitespace_token1, size:1
shift state:12
process version:0, version_count:1, state:12, row:1, col:12
lex_internal state:4, row:1, column:12
  consume character:'t'
  consume character:'o'
  consume character:'t'
  consume character:'o'
```
Is there a guide anywhere to explain how to interpret/read those ?  the first error appears after parsing "X-header-0", then it seems to reparse it successfully. Here is the full grammar if you are motivated https://github.com/teto/tree-sitter-http/blob/4f3f8bee388951ee3f7abaff08430eecafe162aa/grammar.js . Meanwhile I am doing a more thorough search in "discussions" trying to find a similar case.

**Comment by sogaiu** [comment]
FWIW, there is [this](https://github.com/tree-sitter/tree-sitter/issues/1309), but I don't know how immediately helpful that will be (the talks might help a little perhaps).  Have you also seen the `--debug-graph` functionality?  That should lead to a `log.html` file containing SVG which should be viewable in some web browser (forgot to mention that graphviz needs to be installed as I believe the `dot` command is used).

Sorry I don't have much to offer in the way of specifics, but perhaps other folks might be able to help.

**Reply by teto** [reply]
thanks for the links, the wikipedia page about shift-reduce was the most useful for me. That knowledge helps understanding the log but not necessarily why there are conflicts. I've been trying various things so far with little success. I am not sure how newlines are dealt with. It seems like specifying where to expect newlines helps get rid of some errors.

**Comment by sogaiu** [comment]
Don't know if this will be helpful, but from [here](https://github.com/tree-sitter/tree-sitter/issues/1309#issuecomment-894582457), I see a link to the source that emits `detect error`.  Looking at the source: https://github.com/tree-sitter/tree-sitter/blob/fb6bd2cad3778bc8cc4c1eef053d56fba4a5040e/lib/src/parser.c#L1575-L1580
